ext/app/electron/GristApp.ts

```diff
@@ -1,319 +1,116 @@
-      default:
-        await this.openGristFile(filepath).catch(e => this.reportError(e));
-        break;
-    }
-  }
+  private static _instance: GristApp; // The singleton instance.
 
-  private getUrl(docID?: string) {
-    const url = new URL(this.appHost);
-    if (docID) {
-      url.pathname = 'doc/' + encodeURIComponent(docID);
-    }
-    if (this.authMode !== 'none') {
-      url.searchParams.set('electron_key', this.credential);
-    }
-    return url.href;
-  }
+  // This is referenced by create.ts.
+  // TODO: Should we make DocRegistry its own singleton class? (To avoid circular import.)
+  public docRegistry: DocRegistry;
+  private flexServer: FlexServer;
+  private windowManager: WindowManager;
 
-  private async openGristFile(filepath: string, openWith?: {loadURL: (url: string) => Promise<void>}) {
-    const target = await this.normalizePath(filepath);
-    const docsRoot = this.flexServer.docsRoot;
-    const root = await this.normalizePath(docsRoot);
-    console.log("Opening a file", {
-      filepath,
-      target,
-      docsRoot,
-      root,
-    });
+  private constructor() {}
 
-    // Here is our dumb strategy for opening random Grist files on the
-    // file system: just mint a key and soft-link to them. If being
-    // professional, Grist should be watching for external modifications.
-    // Baby steps though.
-    let docId: string|undefined;
-    let maybeDocId: string|undefined;
-    if (!path.relative(root, target).startsWith('..')) {
-      const did = path.basename(target, '.grist');
-      const p = path.join(docsRoot, `${did}.grist`);
-      if (await this.normalizePath(p) === target) {
-        maybeDocId = did;
-      }
-    }
-    const db = this.flexServer.getHomeDBManager();
-    const user = await db.getUserByLogin(process.env.GRIST_DEFAULT_EMAIL as string);
-    if (!user) { throw new Error('cannot find default user'); }
-    const wss = db.unwrapQueryResult(await db.getOrgWorkspaces({userId: user.id}, 0));
-    for (const doc of wss[0].docs) {
-      if (doc.options?.externalId === target || doc.id === maybeDocId) {
-        docId = doc.id;
-        break;
-      }
+  public static get instance(): GristApp {
+    if (!GristApp._instance) {
+      GristApp._instance = new GristApp();
     }
-    if (!docId) {
-      docId = db.unwrapQueryResult(await db.addDocument({
-        userId: user.id,
-      }, wss[0].id, {
-        name: path.basename(target, '.grist'),
-        options: { externalId: target },
-      }));
-    }
-    const link = path.join(docsRoot, `${docId}.grist`);
-    if (!fse.pathExistsSync(link)) {
-      fse.symlinkSync(target, link);
-    }
-    this.openWindowForDoc(docId, openWith);
+    return GristApp._instance;
   }
 
-  // Returns the last Grist window that was created.
-  private getLastWindow() {
-    let lastWindow = null;
-    for (const win of this.appWindows) {
-      lastWindow = win;
-    }
-    return lastWindow;
+  // TODO: Move this function somewhere else.
```

Comment:
I'm not sure either, so I left it here and added this comment. When this project grows larger, it would be better to keep `GristApp` simple and avoid having it hold too many methods.